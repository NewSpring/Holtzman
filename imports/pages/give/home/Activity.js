// @flow
import React, { Component } from "react";
import { Link } from "react-router";

import SectionHeader from "../../../components/@primitives/UI/section-header";
import LoadingHeader from "../../../components/@primitives/UI/loading/SectionHeader";
import SmallButton from "../../../components/@primitives/UI/buttons/SmallButton";
import ActivityCard from "../../../components/giving/cards/ActivityCard";
import LoadingActivityCard from "../../../components/@primitives/UI/loading/ActivityCard";

import SummaryChart from "./GivingSummary";

const ActivityButton = () =>
  (<SmallButton
    text="See All"
    linkUrl="/give/history"
    className="btn--dark-tertiary flush"
  />);

type IGivingActivity = {
  feed: Object,
};

export class GivingActivity extends Component {
  props: IGivingActivity;

  filterActivity = (data: [Object]): Array<any> => {
    if (!Array.isArray(data)) return [];

    const transactions = [];
    // const accounts = [{}];
    const activityToShow = [];

    // separate feed by transactions and accoutns
    data.forEach((feedItem: Object) => {
      if (typeof feedItem.status !== "undefined") {
        transactions.push(feedItem);
      }
      // else {
      //   accounts.push(feedItem);
      // }
    });

    // XXX TODO: filter accounts by soon expiring

    // accounts.map((account) => {
    //   if (activityToShow.length < 3) {
    //     activityToShow.push(account);
    //   }
    //   return null;
    // });
    transactions.map(transaction => {
      if (transaction.details[0].amount !== 0) {
        if (activityToShow.length < 3) {
          activityToShow.push(transaction);
        }
      }
      return null;
    });

    return activityToShow;
  };

  additionalAmount = (details: Object) => {
    const amount = details.amount === parseInt(details.amount, 10)
      ? details.amount
      : details.amount.toFixed(2);

    return (
      <span>and <strong>${amount}</strong> to <strong>{details.account.name}</strong></span>
    );
  };

  // used by renderActivity to render a transasction card
  renderTransaction = (transaction: Object): any => {
    let message;
    let status;
    let linkText;
    let linkUrl;
    const scheduled = transaction.schedule !== null;

    // show .00 on whole-dollar amounts. Don't show on even dollars
    const amount = transaction.details[0].amount === parseInt(transaction.details[0].amount, 10)
      ? transaction.details[0].amount
      : transaction.details[0].amount.toFixed(2);

    if ((transaction.status === null || transaction.status === "Success" || transaction.status === "Complete") && transaction.details.length) {
      status = "success";
      linkText = "View Contribution";
      linkUrl = `/give/history/${transaction.id}`;
      message = (
        <p>
          Your {scheduled ? "scheduled " : ""}contribution of <strong>${amount} </strong>
          to <strong>{transaction.details[0].account.name} </strong>
          {transaction.details.length > 1 ? this.additionalAmount(transaction.details[1]) : null}
          {" "}was successful.
        </p>
      );
    } else if (transaction.status === "Failed") {
      status = "failed";
      message = (
        <p>
          Your {scheduled ? "scheduled " : ""}contribution to
          <strong> {transaction.details[0].account.name} </strong>
          {transaction.details.length > 1
            ? <span>and<strong> {transaction.details[1].name} </strong></span> : null}
          {" "}was unsuccessful.
          {transaction.statusMessage !== null && transaction.statusMessage !== ""
            ? ` Unfortunately, ${transaction.statusMessage}.` : ""}
        </p>
      );
    } else if (transaction.status === "Pending") {
      status = "success";
      message = (
        <p>
          Your {scheduled ? "scheduled " : ""}contribution to
          <strong> {transaction.details[0].account.name} </strong>
          {transaction.details.length > 1
            ? <span>and<strong> {transaction.details[1].name} </strong></span> : null}
          {" "}is <strong>pending</strong>.
        </p>
      );
    } else {
      return null;
    }

    return (
      <ActivityCard
        key={`${transaction.id.slice(0, -8)}_${transaction.details[0].account.name}`}
        status={status}
        date={transaction.date}
        message={message}
        linkText={linkText}
        linkUrl={linkUrl}
      />
    );
  };

  renderExpiringAccount = (account: Object): any =>
    (<ActivityCard
      key={account.id}
      status=""
      message={`Your saved payment ${account.name} is expiring soon.`}
      linkText={"Update it Now"}
      linkUrl={"/"}
      date={"2016-12-01"}
    />)
  ;

  renderActivity = (feedItems: ?Array<Object>): any => {
    if (Array.isArray(feedItems) && feedItems.length > 0) {
      return feedItems.map((feedItem: Object) => {
        if (feedItem.__typename === "Transaction") { //eslint-disable-line
          return this.renderTransaction(feedItem);
        }
        return this.renderExpiringAccount(feedItem);
      });
    }
    return null;
  };

  render() {
    const wrapper = "soft-half-sides soft-double-sides@lap-and-up";

    if (!this.props.feed || !this.props.feed.userFeed) {
      return (
        <div className={wrapper}>
          <LoadingHeader />
          <LoadingActivityCard />
        </div>
      );
    }

    const data = this.filterActivity(this.props.feed.userFeed);
    if (!Array.isArray(data) && data.length === 0) {
      return (
        <div className={wrapper}>
          <SectionHeader title="Activity" link={<div />} />
          <div className="card">
            <div className="card__item soft">
              <h4 className="text-dark-primary">
                {"You don't have any activity to show"}
              </h4>
              <p>
                {`This section is to keep you up to date on your recent online giving
                activity. It appears as if you haven't given online before.`}
              </p>
              <p>
                {`If you believe this is an error and you would like a member of our
                customer support team to contact you, click `}
                <a
                  target="_blank"
                  rel="noopener noreferrer"
                  href="//rock.newspring.cc/workflows/152?Topic=Stewardship"
                >
                  here
                </a>.
              </p>
              <Link to="/give/now" className="btn one-whole@handheld flush-bottom">
                Give Now
              </Link>
            </div>
          </div>
        </div>
      );
    }

    return (
      <div className={wrapper}>
        <SectionHeader title="Activity" link={<ActivityButton />} />
        <div className="hard">
          {this.renderActivity(data)}
        </div>
        <SummaryChart />
      </div>
    );
  }
}

export default GivingActivity;
