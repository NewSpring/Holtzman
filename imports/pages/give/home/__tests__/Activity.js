
import { GivingActivity } from "../Activity";
import { mount } from "enzyme";
import { mountToJson } from "enzyme-to-json";

jest.mock("../GivingSummary", () => "GivingSummary");

const mockFeedData = [
  {
    "id": "123",
    "date": "2016-12-01",
    "summary": "Reference Number: 12345",
    "status": "Pending",
    "statusMessage": "Manually set to pending for testing",
    "schedule": null,
    "details": [{"amount": 1,"account": {"name": "General Fund"}}]
  },
  {
    "id": "456",
    "date": "2016-11-01",
    "summary": "Reference Number: 2355654",
    "status": "Complete",
    "statusMessage": "Manually set to failed for testing",
    "schedule": null,
    "details": [{"amount": 2,"account": {"name": "General Fund"}}]
  },
  {
    "id": "457",
    "date": "2016-11-01",
    "summary": "Reference Number: 2355654",
    "status": "Failed",
    "statusMessage": "Manually set to failed for testing",
    "schedule": null,
    "details": [{"amount": 2,"account": {"name": "General Fund"}}]
  },
  {
    "id": "789",
    "date": "2016-10-03",
    "summary": "Reference Number: 76543",
    "status": null,
    "statusMessage": null,
    "schedule": null,
    "details": [{"amount": 3,"account": {"name": "General Fund"}}, {"amount": 2,"account": {"name": "Harambe Fund"}}]
  },
  {
    "id": "890",
    "date": "2016-09-03",
    "summary": "Reference Number: 89098",
    "status": null,
    "statusMessage": null,
    "schedule": null,
    "details": [{"amount": 0,"account": {"name": "General Fund"}}, {"amount": 0,"account": {"name": "Harambe Fund"}}]
  },
  {
    "id": "111",
    "name": "Harambe's Card",
    "expirationYear": null,
    "expirationMonth": null
  }
];

const generateComponent = (additionalProps) =>
  <GivingActivity {...additionalProps} />;

describe("GivingActivity", () => {
  it("should render with minimal props and display no activity placeholder", () => {
    const component = mount(generateComponent());
    expect(mountToJson(component)).toMatchSnapshot();
  });

  /**
   * this should be all that's needed, since the component only renders activityCards.
   * We just need to check for the correct props passed to the components.
   * Since the mock data covers all the possible statuses of transactions, this should
   * be fine.
   */
  it("should render with data", () => {
    const component = mount(generateComponent({
      feed: { userFeed: mockFeedData },
    }));
    expect(mountToJson(component)).toMatchSnapshot();
  });
});
