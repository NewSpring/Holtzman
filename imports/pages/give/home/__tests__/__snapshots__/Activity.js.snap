exports[`GivingActivity should render with data 1`] = `
<GivingActivity
  feed={
    Object {
      "userFeed": Array [
        Object {
          "date": "2016-12-01",
          "details": Array [
            Object {
              "account": Object {
                "name": "General Fund",
              },
              "amount": 1,
            },
          ],
          "id": "123",
          "schedule": null,
          "status": "Pending",
          "statusMessage": "Manually set to pending for testing",
          "summary": "Reference Number: 12345",
        },
        Object {
          "date": "2016-11-01",
          "details": Array [
            Object {
              "account": Object {
                "name": "General Fund",
              },
              "amount": 2,
            },
          ],
          "id": "456",
          "schedule": null,
          "status": "Failed",
          "statusMessage": "Manually set to failed for testing",
          "summary": "Reference Number: 2355654",
        },
        Object {
          "date": "2016-10-03",
          "details": Array [
            Object {
              "account": Object {
                "name": "General Fund",
              },
              "amount": 3,
            },
          ],
          "id": "789",
          "schedule": null,
          "status": null,
          "statusMessage": null,
          "summary": "Reference Number: 76543",
        },
        Object {
          "expirationMonth": null,
          "expirationYear": null,
          "id": "111",
          "name": "Harambe\'s Card",
        },
      ],
    }
  }>
  <div
    className="soft-half hard-top">
    <Activity
      date="2016-12-01"
      linkText="Update it Now"
      linkUrl="/"
      message="Your saved payment undefined is expiring soon."
      status="">
      <div
        className="card">
        <div
          className="card__item soft text-left soft-bottom rounded background--dark-primary text-light-primary">
          <div
            className="floating text-left">
            <i
              className="soft-half-right">
              <Error
                fill="#FFFFFF"
                height="31px"
                width="30px">
                <svg
                  height="31px"
                  version="1.1"
                  viewBox="0 0 60 61"
                  width="30px">
                  <g
                    fill="none"
                    fillRule="evenodd"
                    stroke="none"
                    strokeWidth="1">
                    <g
                      fill="#FFFFFF"
                      transform="translate(-90.000000, -163.000000)">
                      <g
                        transform="translate(60.000000, 0.000000)">
                        <g
                          transform="translate(30.000000, 163.000000)">
                          <path
                            d="M30,54.33 L30,54.33 C43.254834,54.33 54,43.584834 54,30.33 C54,17.075166 43.254834,6.33 30,6.33 C16.745166,6.33 6,17.075166 6,30.33 C6,43.584834 16.745166,54.33 30,54.33 L30,54.33 Z M30,60.33 L30,60.33 C13.4314575,60.33 0,46.8985425 0,30.33 C0,13.7614575 13.4314575,0.33 30,0.33 C46.5685425,0.33 60,13.7614575 60,30.33 C60,46.8985425 46.5685425,60.33 30,60.33 L30,60.33 Z"
                            id="Shape" />
                          <path
                            d="M28.784,33.438 L28.784,33.438 L27.9222343,33.438 L27.7949654,32.5856839 L26.7389654,25.5136839 L26.728,25.366 L26.728,16.502 L26.728,15.502 L27.728,15.502 L32.272,15.502 L33.272,15.502 L33.272,16.502 L33.272,25.366 L33.2610346,25.5136839 L32.2050346,32.5856839 L32.0777657,33.438 L31.216,33.438 L28.784,33.438 Z M26.504,41.662 L26.504,41.662 C26.504,39.6282997 27.9530061,38.166 30,38.166 C32.0469939,38.166 33.496,39.6282997 33.496,41.662 C33.496,43.6957003 32.0469939,45.158 30,45.158 C27.9530061,45.158 26.504,43.6957003 26.504,41.662 L26.504,41.662 Z"
                            id="Shape" />
                        </g>
                      </g>
                    </g>
                  </g>
                </svg>
              </Error>
            </i>
            <h5
              className="display-inline-block floating__item soft-half-bottom">
              Dec 1, 2016
            </h5>
          </div>
          <p>
            Your saved payment undefined is expiring soon.
          </p>
          <Link
            className="text-light-primary plain"
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/">
            <a
              className="text-light-primary plain"
              onClick={[Function]}
              style={Object {}}>
              <h5
                className="display-inline-block">
                Update it Now
              </h5>
              <span
                className="icon-arrow-next soft-half-left" />
            </a>
          </Link>
        </div>
      </div>
    </Activity>
    <Activity
      date="2016-12-01"
      linkText="Update it Now"
      linkUrl="/"
      message="Your saved payment undefined is expiring soon."
      status="">
      <div
        className="card">
        <div
          className="card__item soft text-left soft-bottom rounded background--dark-primary text-light-primary">
          <div
            className="floating text-left">
            <i
              className="soft-half-right">
              <Error
                fill="#FFFFFF"
                height="31px"
                width="30px">
                <svg
                  height="31px"
                  version="1.1"
                  viewBox="0 0 60 61"
                  width="30px">
                  <g
                    fill="none"
                    fillRule="evenodd"
                    stroke="none"
                    strokeWidth="1">
                    <g
                      fill="#FFFFFF"
                      transform="translate(-90.000000, -163.000000)">
                      <g
                        transform="translate(60.000000, 0.000000)">
                        <g
                          transform="translate(30.000000, 163.000000)">
                          <path
                            d="M30,54.33 L30,54.33 C43.254834,54.33 54,43.584834 54,30.33 C54,17.075166 43.254834,6.33 30,6.33 C16.745166,6.33 6,17.075166 6,30.33 C6,43.584834 16.745166,54.33 30,54.33 L30,54.33 Z M30,60.33 L30,60.33 C13.4314575,60.33 0,46.8985425 0,30.33 C0,13.7614575 13.4314575,0.33 30,0.33 C46.5685425,0.33 60,13.7614575 60,30.33 C60,46.8985425 46.5685425,60.33 30,60.33 L30,60.33 Z"
                            id="Shape" />
                          <path
                            d="M28.784,33.438 L28.784,33.438 L27.9222343,33.438 L27.7949654,32.5856839 L26.7389654,25.5136839 L26.728,25.366 L26.728,16.502 L26.728,15.502 L27.728,15.502 L32.272,15.502 L33.272,15.502 L33.272,16.502 L33.272,25.366 L33.2610346,25.5136839 L32.2050346,32.5856839 L32.0777657,33.438 L31.216,33.438 L28.784,33.438 Z M26.504,41.662 L26.504,41.662 C26.504,39.6282997 27.9530061,38.166 30,38.166 C32.0469939,38.166 33.496,39.6282997 33.496,41.662 C33.496,43.6957003 32.0469939,45.158 30,45.158 C27.9530061,45.158 26.504,43.6957003 26.504,41.662 L26.504,41.662 Z"
                            id="Shape" />
                        </g>
                      </g>
                    </g>
                  </g>
                </svg>
              </Error>
            </i>
            <h5
              className="display-inline-block floating__item soft-half-bottom">
              Dec 1, 2016
            </h5>
          </div>
          <p>
            Your saved payment undefined is expiring soon.
          </p>
          <Link
            className="text-light-primary plain"
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/">
            <a
              className="text-light-primary plain"
              onClick={[Function]}
              style={Object {}}>
              <h5
                className="display-inline-block">
                Update it Now
              </h5>
              <span
                className="icon-arrow-next soft-half-left" />
            </a>
          </Link>
        </div>
      </div>
    </Activity>
    <Activity
      date="2016-12-01"
      linkText="Update it Now"
      linkUrl="/"
      message="Your saved payment undefined is expiring soon."
      status="">
      <div
        className="card">
        <div
          className="card__item soft text-left soft-bottom rounded background--dark-primary text-light-primary">
          <div
            className="floating text-left">
            <i
              className="soft-half-right">
              <Error
                fill="#FFFFFF"
                height="31px"
                width="30px">
                <svg
                  height="31px"
                  version="1.1"
                  viewBox="0 0 60 61"
                  width="30px">
                  <g
                    fill="none"
                    fillRule="evenodd"
                    stroke="none"
                    strokeWidth="1">
                    <g
                      fill="#FFFFFF"
                      transform="translate(-90.000000, -163.000000)">
                      <g
                        transform="translate(60.000000, 0.000000)">
                        <g
                          transform="translate(30.000000, 163.000000)">
                          <path
                            d="M30,54.33 L30,54.33 C43.254834,54.33 54,43.584834 54,30.33 C54,17.075166 43.254834,6.33 30,6.33 C16.745166,6.33 6,17.075166 6,30.33 C6,43.584834 16.745166,54.33 30,54.33 L30,54.33 Z M30,60.33 L30,60.33 C13.4314575,60.33 0,46.8985425 0,30.33 C0,13.7614575 13.4314575,0.33 30,0.33 C46.5685425,0.33 60,13.7614575 60,30.33 C60,46.8985425 46.5685425,60.33 30,60.33 L30,60.33 Z"
                            id="Shape" />
                          <path
                            d="M28.784,33.438 L28.784,33.438 L27.9222343,33.438 L27.7949654,32.5856839 L26.7389654,25.5136839 L26.728,25.366 L26.728,16.502 L26.728,15.502 L27.728,15.502 L32.272,15.502 L33.272,15.502 L33.272,16.502 L33.272,25.366 L33.2610346,25.5136839 L32.2050346,32.5856839 L32.0777657,33.438 L31.216,33.438 L28.784,33.438 Z M26.504,41.662 L26.504,41.662 C26.504,39.6282997 27.9530061,38.166 30,38.166 C32.0469939,38.166 33.496,39.6282997 33.496,41.662 C33.496,43.6957003 32.0469939,45.158 30,45.158 C27.9530061,45.158 26.504,43.6957003 26.504,41.662 L26.504,41.662 Z"
                            id="Shape" />
                        </g>
                      </g>
                    </g>
                  </g>
                </svg>
              </Error>
            </i>
            <h5
              className="display-inline-block floating__item soft-half-bottom">
              Dec 1, 2016
            </h5>
          </div>
          <p>
            Your saved payment undefined is expiring soon.
          </p>
          <Link
            className="text-light-primary plain"
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/">
            <a
              className="text-light-primary plain"
              onClick={[Function]}
              style={Object {}}>
              <h5
                className="display-inline-block">
                Update it Now
              </h5>
              <span
                className="icon-arrow-next soft-half-left" />
            </a>
          </Link>
        </div>
      </div>
    </Activity>
  </div>
</GivingActivity>
`;

exports[`GivingActivity should render with minimal props 1`] = `<GivingActivity />`;
